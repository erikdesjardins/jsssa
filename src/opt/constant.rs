use std::collections::HashMap;
use std::f64;

use crate::collections::ZeroOneMany::{self, Many, One, Zero};
use crate::ir;
use crate::ir::traverse::Folder;

/// Constant propagation / precompute.
///
/// Does not profit from multiple passes.
/// Does not profit from DCE running first; may create opportunities for DCE.
#[derive(Default)]
pub struct ConstProp {
    shallow_values: HashMap<ir::WeakRef<ir::Ssa>, ir::Expr>,
}

impl ConstProp {
    fn maybe_shallow_cache_expr(&mut self, ref_: &ir::Ref<ir::Ssa>, expr: &ir::Expr) {
        // avoid cloning refs, which wastes time on refcounts and make tracing ref drops harder
        // and CERTAINLY avoid deep cloning blocks
        let shallow_clone = match expr {
            ir::Expr::Bool { value } => ir::Expr::Bool { value: *value },
            ir::Expr::Number { value } => ir::Expr::Number { value: *value },
            ir::Expr::String { value } => ir::Expr::String {
                value: value.clone(),
            },
            ir::Expr::Null => ir::Expr::Null,
            ir::Expr::Undefined => ir::Expr::Undefined,
            ir::Expr::This => ir::Expr::This,
            // avoid cloning refs inside array
            ir::Expr::Array { elems } if elems.is_empty() => ir::Expr::Array { elems: vec![] },
            // avoid cloning refs inside object
            ir::Expr::Object { props } if props.is_empty() => ir::Expr::Object { props: vec![] },
            ir::Expr::RegExp { regex, flags } => ir::Expr::RegExp {
                regex: regex.clone(),
                flags: flags.clone(),
            },
            // avoid cloning function body
            ir::Expr::Function { kind, body: _ } => ir::Expr::Function {
                kind: kind.clone(),
                body: ir::Block(vec![]),
            },
            ir::Expr::CurrentFunction => ir::Expr::CurrentFunction,
            ir::Expr::Argument { index } => ir::Expr::Argument { index: *index },
            _ => return,
        };

        self.shallow_values.insert(ref_.weak(), shallow_clone);
    }
}

impl Folder for ConstProp {
    type Output = ZeroOneMany<ir::Stmt>;

    #[rustfmt::skip]
    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        use ir::BinaryOp::*;
        use ir::Expr::*;
        use ir::UnaryOp::*;
        use ir::F64;

        #[allow(clippy::cast_lossless, clippy::float_cmp)]
        match stmt {
            ir::Stmt::Expr {
                target,
                expr,
            } => {
                let expr = match expr {
                    ReadGlobal { ref source } if source == "NaN" => Number { value: F64(f64::NAN) },
                    ReadGlobal { ref source } if source == "undefined" => Undefined,
                    Unary {
                        ref op,
                        ref val,
                    } => match self.shallow_values.get(&val.weak()) {
                        Some(val_val) => match (op, val_val) {
                            (Plus, Number { value }) => Number { value: *value },
                            (Minus, Number { value }) => Number { value: F64(-value.0) },
                            (Not, Number { value }) => Bool { value: value.0 == 0.0 },
                            (Not, Bool { value }) => Bool { value: !*value },
                            (Tilde, Number { value }) => Number { value: F64(!(value.0 as i32) as f64) },
                            (Typeof, Bool { .. }) => String { value: "boolean".into() },
                            (Typeof, Number { .. }) => String { value: "number".into() },
                            (Typeof, String { .. }) => String { value: "string".into() },
                            (Typeof, Null) => String { value: "object".into() },
                            (Typeof, Undefined) => String { value: "undefined".into() },
                            (Typeof, Array { .. })
                            | (Typeof, Object { .. })
                            | (Typeof, RegExp { .. }) => String { value: "object".into() },
                            (Typeof, Function { .. })
                            | (Typeof, CurrentFunction) => String { value: "function".into() },
                            (Void, _) => Undefined,
                            _ => expr,
                        },
                        None => expr,
                    },
                    Binary {
                        ref op,
                        ref left,
                        ref right,
                    } => match (self.shallow_values.get(&left.weak()), self.shallow_values.get(&right.weak())) {
                        (Some(left_val), Some(right_val)) => match (op, left_val, right_val) {
                            (EqEq, _, _)
                            | (StrictEq, _, _) if left == right => Bool { value: true },
                            (EqEq, Bool { value: a }, Bool { value: b })
                            | (StrictEq, Bool { value: a }, Bool { value: b }) => Bool { value: a == b },
                            (EqEq, Number { value: a }, Number { value: b })
                            | (StrictEq, Number { value: a }, Number { value: b }) => Bool { value: a.0 == b.0 },
                            (NotEq, Bool { value: a }, Bool { value: b })
                            | (NotStrictEq, Bool { value: a }, Bool { value: b }) => Bool { value: a != b },
                            (NotEq, Number { value: a }, Number { value: b })
                            | (NotStrictEq, Number { value: a }, Number { value: b }) => Bool { value: a.0 != b.0 },
                            (Lt, Number { value: a }, Number { value: b }) => Bool { value: a.0 < b.0 },
                            (LtEq, Number { value: a }, Number { value: b }) => Bool { value: a.0 <= b.0 },
                            (Gt, Number { value: a }, Number { value: b }) => Bool { value: a.0 > b.0 },
                            (GtEq, Number { value: a }, Number { value: b }) => Bool { value: a.0 >= b.0 },
                            (ShiftLeft, Number { value: a }, Number { value: b }) => Number { value: F64(((a.0 as i32) << b.0 as i32) as f64) },
                            (ShiftRight, Number { value: a }, Number { value: b }) => Number { value: F64(((a.0 as i32) >> b.0 as i32) as f64) },
                            (ShiftRightZero, Number { value: a }, Number { value: b }) => Number { value: F64(((a.0 as i32 as u32) >> b.0 as i32) as f64) },
                            (Add, Number { value: a }, Number { value: b }) => Number { value: F64(a.0 + b.0) },
                            (Sub, Number { value: a }, Number { value: b }) => Number { value: F64(a.0 - b.0) },
                            (Mul, Number { value: a }, Number { value: b }) => Number { value: F64(a.0 * b.0) },
                            (Div, Number { value: a }, Number { value: b }) => Number { value: F64(a.0 / b.0) },
                            (Mod, Number { value: a }, Number { value: b }) => Number { value: F64(a.0 % b.0) },
                            (BitOr, Number { value: a }, Number { value: b }) => Number { value: F64((a.0 as i32 | b.0 as i32) as f64) },
                            (BitXor, Number { value: a }, Number { value: b }) => Number { value: F64((a.0 as i32 ^ b.0 as i32) as f64) },
                            (BitAnd, Number { value: a }, Number { value: b }) => Number { value: F64((a.0 as i32 & b.0 as i32) as f64) },
                            (Exp, Number { value: a }, Number { value: b }) => Number { value: F64(a.0.powf(b.0)) },
                            (Add, String { value: a }, String { value: b }) => String { value: (a.to_string() + b).into() },
                            _ => expr,
                        },
                        _ => expr,
                    },
                    _ => expr,
                };

                self.maybe_shallow_cache_expr(&target, &expr);

                One(ir::Stmt::Expr { target, expr })
            }
            ir::Stmt::ForEach {
                ref kind,
                ref init,
                body: _,
            } => match self.shallow_values.get(&init.weak()) {
                Some(init_val) => match (kind, init_val) {
                    (ir::ForKind::In, Object { props }) if props.is_empty() => Zero,
                    (ir::ForKind::Of, Array { elems }) if elems.is_empty() => Zero,
                    _ => One(stmt),
                },
                None => One(stmt),
            },
            ir::Stmt::IfElse {
                cond,
                cons,
                alt,
            } => match self.shallow_values.get(&cond.weak()) {
                Some(cond_val) => match cond_val {
                    Bool { value: true }
                    | Array { .. }
                    | Object { .. }
                    | RegExp { .. }
                    | Function { .. } => Many(cons.0),
                    Bool { value: false }
                    | Null
                    | Undefined => Many(alt.0),
                    Number { value } => if value.0 == 0. || value.0.is_nan() { Many(alt.0) } else { Many(cons.0) },
                    String { value } => if value == "" { Many(alt.0) } else { Many(cons.0) },
                    _ => One(ir::Stmt::IfElse { cond, cons, alt }),
                },
                None => One(ir::Stmt::IfElse { cond, cons, alt }),
            },
            _ => One(stmt),
        }
    }
}
