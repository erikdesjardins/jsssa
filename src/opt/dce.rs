use crate::ir;
use crate::ir::fold::Folder;

#[derive(Default)]
pub struct Dce {
    dropping_after_jump: bool,
}

impl Folder for Dce {
    type Output = Option<ir::Stmt>;

    fn wrap_scope<R>(&mut self, enter: impl FnOnce(&mut Self) -> R) -> R {
        let r = enter(self);
        // stop dropping when we leave the current scope
        self.dropping_after_jump = false;
        r
    }

    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        if self.dropping_after_jump {
            return None;
        }

        match stmt {
            ir::Stmt::Expr {
                ref target,
                ref expr,
            } if target.used().is_never() => match expr {
                ir::Expr::Function { .. }
                | ir::Expr::Bool { .. }
                | ir::Expr::Number { .. }
                | ir::Expr::String { .. }
                | ir::Expr::Null
                | ir::Expr::Undefined
                | ir::Expr::This
                | ir::Expr::Read { .. }
                | ir::Expr::ReadMutable { .. }
                | ir::Expr::Array { .. }
                | ir::Expr::Object { .. }
                | ir::Expr::RegExp { .. }
                | ir::Expr::Unary { .. }
                | ir::Expr::Binary { .. }
                | ir::Expr::CurrentFunction
                | ir::Expr::Argument { .. } => None,
                ir::Expr::ReadGlobal { .. }
                | ir::Expr::ReadMember { .. }
                | ir::Expr::Delete { .. }
                | ir::Expr::Yield { .. }
                | ir::Expr::Await { .. }
                | ir::Expr::Call { .. } => Some(stmt),
            },
            ir::Stmt::DeclareMutable { ref target, val: _ } if target.used().is_never() => None,
            ir::Stmt::Return { .. }
            | ir::Stmt::Throw { .. }
            | ir::Stmt::Break
            | ir::Stmt::Continue => {
                self.dropping_after_jump = true;
                Some(stmt)
            }
            ir::Stmt::IfElse {
                cond: _,
                ref cons,
                ref alt,
            } if cons.0.is_empty() && alt.0.is_empty() => None,
            ir::Stmt::Try {
                ref body,
                catch: _,
                ref finally,
            } if body.0.is_empty() && finally.0.is_empty() => None,
            _ => Some(stmt),
        }
    }
}
