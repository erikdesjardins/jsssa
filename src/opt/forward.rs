use std::collections::HashMap;
use std::iter;

use crate::ir;
use crate::ir::traverse::{visit_with, Folder, ScopeTy};

/// Forward `ir::Expr::Read` to the source SSA ref.
///
/// Does not profit from multiple passes.
/// Does not profit from DCE running first; may create opportunities for DCE.
#[derive(Default)]
pub struct Reads {
    ssa_remappings: HashMap<ir::WeakRef<ir::Ssa>, ir::Ref<ir::Ssa>>,
}

impl Reads {
    fn remap(&self, ref_: ir::Ref<ir::Ssa>) -> ir::Ref<ir::Ssa> {
        match self.ssa_remappings.get(&ref_.weak()) {
            Some(orig_ref) => orig_ref.clone(),
            None => ref_,
        }
    }
}

impl Folder for Reads {
    type Output = iter::Once<ir::Stmt>;

    fn wrap_scope<R>(
        &mut self,
        ty: &ScopeTy,
        block: ir::Block,
        enter: impl FnOnce(&mut Self, ir::Block) -> R,
    ) -> R {
        if let ScopeTy::Toplevel = ty {
            visit_with(&block, |stmt: &ir::Stmt| match stmt {
                ir::Stmt::Expr {
                    target,
                    expr: ir::Expr::Read { source },
                } => match self.ssa_remappings.get(&source.weak()) {
                    Some(orig_ref) => {
                        self.ssa_remappings.insert(target.weak(), orig_ref.clone());
                    }
                    None => {
                        self.ssa_remappings.insert(target.weak(), source.clone());
                    }
                },
                _ => {}
            });
        }

        enter(self, block)
    }

    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        iter::once(match stmt {
            ir::Stmt::Expr { target, expr } => ir::Stmt::Expr {
                target,
                expr: match expr {
                    ir::Expr::Read { source } => ir::Expr::Read {
                        source: self.remap(source),
                    },
                    ir::Expr::ReadMember { obj, prop } => ir::Expr::ReadMember {
                        obj: self.remap(obj),
                        prop: self.remap(prop),
                    },
                    ir::Expr::Array { elems } => ir::Expr::Array {
                        elems: elems
                            .into_iter()
                            .map(|opt| opt.map(|(kind, ele)| (kind, self.remap(ele))))
                            .collect(),
                    },
                    ir::Expr::Object { props } => ir::Expr::Object {
                        props: props
                            .into_iter()
                            .map(|(kind, obj, prop)| (kind, self.remap(obj), self.remap(prop)))
                            .collect(),
                    },
                    ir::Expr::Unary { op, val } => ir::Expr::Unary {
                        op,
                        val: self.remap(val),
                    },
                    ir::Expr::Binary { op, left, right } => ir::Expr::Binary {
                        op,
                        left: self.remap(left),
                        right: self.remap(right),
                    },
                    ir::Expr::Delete { obj, prop } => ir::Expr::Delete {
                        obj: self.remap(obj),
                        prop: self.remap(prop),
                    },
                    ir::Expr::Yield { kind, val } => ir::Expr::Yield {
                        kind,
                        val: self.remap(val),
                    },
                    ir::Expr::Await { val } => ir::Expr::Await {
                        val: self.remap(val),
                    },
                    ir::Expr::Call { kind, func, args } => ir::Expr::Call {
                        kind,
                        func: self.remap(func),
                        args: args
                            .into_iter()
                            .map(|(kind, arg)| (kind, self.remap(arg)))
                            .collect(),
                    },
                    ir::Expr::Bool { .. }
                    | ir::Expr::Number { .. }
                    | ir::Expr::String { .. }
                    | ir::Expr::Null
                    | ir::Expr::Undefined
                    | ir::Expr::This
                    | ir::Expr::ReadMutable { .. }
                    | ir::Expr::ReadGlobal { .. }
                    | ir::Expr::RegExp { .. }
                    | ir::Expr::Function { .. }
                    | ir::Expr::CurrentFunction
                    | ir::Expr::Argument { .. } => expr,
                },
            },
            ir::Stmt::DeclareMutable { target, val } => ir::Stmt::DeclareMutable {
                target,
                val: self.remap(val),
            },
            ir::Stmt::WriteMutable { target, val } => ir::Stmt::WriteMutable {
                target,
                val: self.remap(val),
            },
            ir::Stmt::WriteGlobal { target, val } => ir::Stmt::WriteGlobal {
                target,
                val: self.remap(val),
            },
            ir::Stmt::WriteMember { obj, prop, val } => ir::Stmt::WriteMember {
                obj: self.remap(obj),
                prop: self.remap(prop),
                val: self.remap(val),
            },
            ir::Stmt::Return { val } => ir::Stmt::Return {
                val: self.remap(val),
            },
            ir::Stmt::Throw { val } => ir::Stmt::Throw {
                val: self.remap(val),
            },
            ir::Stmt::ForEach { kind, init, body } => ir::Stmt::ForEach {
                kind,
                init: self.remap(init),
                body,
            },
            ir::Stmt::IfElse { cond, cons, alt } => ir::Stmt::IfElse {
                cond: self.remap(cond),
                cons,
                alt,
            },
            ir::Stmt::Switch { discr, body } => ir::Stmt::Switch {
                discr: self.remap(discr),
                body,
            },
            ir::Stmt::Break { .. }
            | ir::Stmt::Continue { .. }
            | ir::Stmt::Debugger
            | ir::Stmt::Label { .. }
            | ir::Stmt::Loop { .. }
            | ir::Stmt::SwitchCase { .. }
            | ir::Stmt::Try { .. } => stmt,
        })
    }
}
