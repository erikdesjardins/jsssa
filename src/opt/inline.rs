use std::collections::{HashMap, HashSet};
use std::iter;
use std::mem;

use crate::collections::ZeroOneMany::{self, Many, One, Zero};
use crate::ir;
use crate::ir::traverse::{Folder, RunVisitor, ScopeTy, Visitor};

/// Inline functions called a single time.
///
/// Does not profit from multiple passes.
/// May profit from DCE running first; may create opportunities for DCE.
/// May create opportunities for _everything_; in particular, read forwarding.
#[derive(Debug, Default)]
pub struct Inline {
    fns_to_inline: HashSet<ir::WeakRef<ir::Ssa>>,
    fn_bodies: HashMap<ir::WeakRef<ir::Ssa>, ir::Block>,
}

impl Drop for Inline {
    fn drop(&mut self) {
        assert!(
            self.fn_bodies.is_empty(),
            "should have inlined fns: {:?}",
            self.fn_bodies
        );
    }
}

#[derive(Debug, Default)]
struct CollectFnCallInfo<'a> {
    fns_to_inline: HashSet<&'a ir::Ref<ir::Ssa>>,
    fn_def_is_good: HashSet<&'a ir::Ref<ir::Ssa>>,
    about_to_enter_arrow_fn: bool,
    about_to_enter_fn: Option<&'a ir::Ref<ir::Ssa>>,
    current_fn: Option<&'a ir::Ref<ir::Ssa>>,
    depth_in_current_fn: u32,
    used_this: bool,
}

impl<'a> Visitor<'a> for CollectFnCallInfo<'a> {
    fn wrap_scope<R>(
        &mut self,
        ty: &ScopeTy,
        block: &'a ir::Block,
        enter: impl FnOnce(&mut Self, &'a ir::Block) -> R,
    ) -> R {
        match ty {
            ScopeTy::Function => {
                let is_arrow_fn = mem::replace(&mut self.about_to_enter_arrow_fn, false);
                let mut inner = Self::default();
                mem::swap(&mut inner.fns_to_inline, &mut self.fns_to_inline);
                mem::swap(&mut inner.fn_def_is_good, &mut self.fn_def_is_good);
                inner.current_fn = self.about_to_enter_fn.take();
                let r = enter(&mut inner, block);
                mem::swap(&mut inner.fns_to_inline, &mut self.fns_to_inline);
                mem::swap(&mut inner.fn_def_is_good, &mut self.fn_def_is_good);
                // if fn hasn't been invalidated, it's good
                self.fn_def_is_good.extend(inner.current_fn);
                // propagate `this` from arrow functions
                if is_arrow_fn && inner.used_this {
                    self.current_fn = None;
                    self.used_this = true;
                }
                r
            }
            ScopeTy::Normal | ScopeTy::Toplevel | ScopeTy::Nonlinear => {
                self.depth_in_current_fn += 1;
                let r = enter(self, block);
                self.depth_in_current_fn -= 1;
                r
            }
        }
    }

    fn visit(&mut self, stmt: &'a ir::Stmt) {
        match stmt {
            ir::Stmt::Expr { target, expr } => match expr {
                ir::Expr::Function {
                    kind:
                        ir::FnKind::Func {
                            is_async: false,
                            is_generator: false,
                        },
                    body: _,
                } if target.used().is_once() => {
                    // fn is simple, used once: def is good if its body is good
                    self.about_to_enter_fn = Some(target);
                }
                ir::Expr::Function { kind, body: _ } => {
                    self.about_to_enter_arrow_fn = true;
                    match kind {
                        ir::FnKind::Arrow { is_async: false } if target.used().is_once() => {
                            // fn is simple, used once: def is good if its body is good
                            self.about_to_enter_fn = Some(target);
                        }
                        _ => {}
                    }
                }
                ir::Expr::Call {
                    kind: ir::CallKind::Call,
                    base,
                    prop: None,
                    args,
                } if self.fn_def_is_good.contains(base) => {
                    if args.iter().all(|(kind, _)| match kind {
                        ir::EleKind::Single => true,
                        ir::EleKind::Spread => false,
                    }) {
                        // fn def is good, and call is simple: mark for inlining
                        self.fns_to_inline.insert(base);
                    }
                }
                ir::Expr::This => {
                    // fn body uses `this`: invalidate
                    self.current_fn = None;
                    self.used_this = true;
                }
                ir::Expr::CurrentFunction => {
                    // fn is possibly recursive: invalidate
                    self.current_fn = None;
                }
                _ => {}
            },
            ir::Stmt::Return { .. } if self.depth_in_current_fn > 0 => {
                // return statement not at top level: invalidate
                self.current_fn = None;
            }
            _ => {}
        }
    }
}

impl Folder for Inline {
    type Output = ZeroOneMany<ir::Stmt>;

    fn wrap_scope<R>(
        &mut self,
        ty: &ScopeTy,
        block: ir::Block,
        enter: impl FnOnce(&mut Self, ir::Block) -> R,
    ) -> R {
        if let ScopeTy::Toplevel = ty {
            let mut collector = CollectFnCallInfo::default();
            collector.run_visitor(&block);
            self.fns_to_inline = collector
                .fns_to_inline
                .into_iter()
                .map(ir::Ref::weak)
                .collect();
        }

        enter(self, block)
    }

    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        match stmt {
            ir::Stmt::Expr {
                target,
                expr: ir::Expr::Function { kind, body },
            } => {
                if self.fns_to_inline.contains(&target.weak()) {
                    self.fn_bodies.insert(target.weak(), body);
                    Zero
                } else {
                    One(ir::Stmt::Expr {
                        target,
                        expr: ir::Expr::Function { kind, body },
                    })
                }
            }
            ir::Stmt::Expr {
                target,
                expr:
                    ir::Expr::Call {
                        kind: ir::CallKind::Call,
                        base,
                        prop: None,
                        args,
                    },
            } => {
                if let Some(ir::Block(body)) = self.fn_bodies.remove(&base.weak()) {
                    let undef_ref = ir::Ref::new("_mis");
                    let undef = ir::Stmt::Expr {
                        target: undef_ref.clone(),
                        expr: ir::Expr::Undefined,
                    };
                    let mut return_ref = None;
                    let body = body
                        .into_iter()
                        .flat_map(|stmt| match stmt {
                            ir::Stmt::Expr {
                                target: _,
                                expr: ir::Expr::CurrentFunction,
                            } => unreachable!("curfn should have been removed"),
                            ir::Stmt::Expr {
                                target,
                                expr: ir::Expr::Argument { index },
                            } => Some(ir::Stmt::Expr {
                                target,
                                expr: ir::Expr::Read {
                                    source: args
                                        .get(index)
                                        .map(|(_, arg)| arg.clone())
                                        .unwrap_or_else(|| undef_ref.clone()),
                                },
                            }),
                            ir::Stmt::Return { val } => {
                                assert!(return_ref.is_none(), "should only have 1 return");
                                return_ref = Some(val);
                                None
                            }
                            _ => Some(stmt),
                        })
                        .collect::<Vec<_>>();
                    let ret = ir::Stmt::Expr {
                        target,
                        expr: ir::Expr::Read {
                            source: return_ref.unwrap_or(undef_ref),
                        },
                    };
                    Many(
                        iter::once(undef)
                            .chain(body)
                            .chain(iter::once(ret))
                            .collect(),
                    )
                } else {
                    One(ir::Stmt::Expr {
                        target,
                        expr: ir::Expr::Call {
                            kind: ir::CallKind::Call,
                            base,
                            prop: None,
                            args,
                        },
                    })
                }
            }
            _ => One(stmt),
        }
    }
}
