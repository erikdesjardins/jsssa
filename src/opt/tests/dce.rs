use crate::opt::dce;

case!(
    basic,
    |cx| cx.converge::<dce::Dce>("dce"),
    r#"
    1;
    true;
    (function() {});
    [];
    ({});
"#,
@"<empty>");

case!(
    basic_bail,
    |cx| cx.converge::<dce::Dce>("dce"),
    r#"
    x;
    const foo;
    foo.bar;
    delete foo.bar;
    foo();
    (function* baz() {
        yield;
    })();
    (async function baz2() {
        await 1;
    })();
"#,
@r###"
<dead> = <global x>
_ini = <void>
foo <= _ini
_obj = *foo
_prp = "bar"
<dead> = _obj[_prp]
_obj$1 = *foo
_prp$1 = "bar"
<dead> = <delete> _obj$1[_prp$1]
_fun = *foo
<dead> = _fun()
_fun$1 = <function generator>:
    _yld = <void>
    <dead> = <yield> _yld
<dead> = _fun$1()
_fun$2 = <function async>:
    _awa = 1
    <dead> = <await> _awa
<dead> = _fun$2()
"###);

case!(
    bindings,
    |cx| cx.converge::<dce::Dce>("dce"),
    r#"
    var x = 1;
    const y = 1;
    function z() {}
"#,
@r###"
_ini = <void>
z <= _ini
_ini$1 = <void>
x <= _ini$1
_fun = <function>:
    <empty>
z <- _fun
_ini$2 = 1
x <- _ini$2
"###);

case!(
    nested_effects,
    |cx| cx.converge::<dce::Dce>("dce"),
    r#"
    [{ x: call() }];
"#,
@r###"
_fun = <global call>
<dead> = _fun()
"###);

case!(
    drop_after_jumps_1,
    |cx| cx.converge::<dce::Dce>("dce"),
    r#"
    (function() {
        good;
        if (x) {
            good;
            throw 1;
            bad;
        }
        good;
        return 2;
        bad;
    })();
"#,
@r###"
_fun = <function>:
    <dead> = <global good>
    _iff = <global x>
    <if> _iff:
        <dead> = <global good>
        _thr = 1
        <throw> _thr
    <else>:
        <empty>
    <dead> = <global good>
    _ret = 2
    <return> _ret
<dead> = _fun()
"###);

case!(
    drop_after_jumps_2,
    |cx| cx.converge::<dce::Dce>("dce"),
    r#"
    for (;;) {
        good;
        if (x) {
            good;
            continue;
            bad;
        }
        good;
        if (y) {
            good;
            break;
            bad;
        }
        good;
    }
"#,
@r###"
<loop>:
    <dead> = <global good>
    _iff = <global x>
    <if> _iff:
        <dead> = <global good>
        <continue>
    <else>:
        <empty>
    <dead> = <global good>
    _iff$1 = <global y>
    <if> _iff$1:
        <dead> = <global good>
        <break>
    <else>:
        <empty>
    <dead> = <global good>
"###);

case!(
    drop_after_jumps_depth,
    |cx| cx.converge::<dce::Dce>("dce"),
    r#"
    (function() {
        good;
        return 2;
        (function() { bad; })();
        if (x) {
            bad;
        }
        bad;
    })();
"#,
@r###"
_fun = <function>:
    <dead> = <global good>
    _ret = 2
    <return> _ret
<dead> = _fun()
"###);

case!(
    dont_drop_hoisted_fns,
    |cx| cx.converge::<dce::Dce>("dce"),
    r#"
    (function() {
        foo();
        return;
        function foo() {
            log();
        }
    })();
"#,
@r###"
_fun = <function>:
    _ini = <void>
    foo <= _ini
    _fun$1 = <function>:
        _fun$3 = <global log>
        <dead> = _fun$3()
    foo <- _fun$1
    _fun$2 = *foo
    <dead> = _fun$2()
    _ret = <void>
    <return> _ret
<dead> = _fun()
"###);

case!(
    empty_blocks,
    |cx| cx.converge::<dce::Dce>("dce"),
    r#"
    if (x) {} else {}
    try {} catch (e) { bad(e); } finally {}
"#,
@"<dead> = <global x>");

case!(
    dont_drop_after_switch_break,
    |cx| cx.converge::<dce::Dce>("dce"),
    r#"
    switch (x) {
        case 1:
            break;
            drop_me();
        case 2:
            log();
    }
"#,
@r###"
_swi = <global x>
_tst = 1
_tst$1 = 2
<switch> _swi:
    <case> _tst:
    <break>
    <case> _tst$1:
    _fun = <global log>
    <dead> = _fun()
"###);

case!(
    do_not_eliminate_for_in_with_assignments,
    all_passes,
    r#"
    let x = {};
    x.y = 1;
    for (let k in x) log(k);
"#,
@r###"
_ini = {  }
_prp = "y"
_val = 1
_ini[_prp] <- _val
<foreach in> _ini:
    _for = <argument 0>
    _fun = <global log>
    <dead> = _fun(_for)
"###);
