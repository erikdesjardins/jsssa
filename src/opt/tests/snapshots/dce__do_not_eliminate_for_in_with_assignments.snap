---
created: "2019-04-27T22:01:39.644416Z"
creator: insta@0.7.4
source: src/opt/tests/dce.rs
expression: "\n    let x = {};\n    foo();\n    x.y = 1;\n    for (let k in x) log(k);\n"
---
x = {  }
_fun = <global foo>
<dead> = _fun()
_prp = "y"
_val = 1
x[_prp] <- _val
<foreach in> x:
    _for = <argument 0>
    _fun$1 = <global log>
    <dead> = _fun$1(_for)
